### PR TITLE
replace all `unreachable!` and `panic!` calls with `bug!`

### LINKED ISSUE #32: Use rustc's compiler error macros.
Miri should use rustc's compiler error macros to ease the eventual merge into rustc itself.

So, instead of `panic!` or `unimplemented!`, use `span_bug!`, `bug!`, etc. I believe these come from `#[macro_use] extern crate rustc;`.

Prompted by https://github.com/solson/miri/pull/28#discussion_r67719292.

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #32 

I didn't touch the `unimplemented!`, since these probably need to go away by being implememented before moving miri into rustc.

### CODE REVIEW

File: src/bin/miri.rs [hunk 1]
```diff
@@ -44,22 +44,25 @@ impl<'a> CompilerCalls<'a> for MiriCompilerCalls {
             fn extract_str(lit: &syntax::ast::Lit) -> syntax::parse::token::InternedString {
                 match lit.node {
                     syntax::ast::LitKind::Str(ref s, _) => s.clone(),
-                    _ => panic!("attribute values need to be strings"),
+                    _ => bug!("attribute values need to be strings"),
```
Comment (bjorn3) [inline src/bin/miri.rs]: This should actually be a span_err, because it is a miri user fault, not miri progrmmer fault.

Comment (oli-obk) [inline src/bin/miri.rs]: I don't think the parser ever yields something other than a `LitKind::Str`. Rustc will error and abort long before we reach this.

Comment (bjorn3) [inline src/bin/miri.rs]: `#[memory_size=true]` should give LitKind::Bool for the value.

File: src/primval.rs [hunk 2]
```diff
@@ -72,11 +72,11 @@ pub fn binary_op<'tcx>(bin_op: mir::BinOp, left: PrimVal, right: PrimVal) -> Eva
...
-                BitXor => unreachable!(),
-                BitAnd => unreachable!(),
-                BitOr  => unreachable!(),
-                Shl => unreachable!(),
-                Shr => unreachable!(),
+                BitXor => bug!("`^` is not a valid operation on floats"),
+                BitAnd => bug!("`&` is not a valid operation on floats"),
+                BitOr  => bug!("`|` is not a valid operation on floats"),
+                Shl => bug!("`<<` is not a valid operation on floats"),
+                Shr => bug!("`>>` is not a valid operation on floats"),
```
Comment (eddyb) [inline src/primval.rs]: These (and other places) could be a single match arm.

File: src/bin/miri.rs (hunk 1 above)
Comment (solson) [inline src/bin/miri.rs]: @oli-obk It turns out this changed in a recent nightly. Beta still bails on non-string literals in attributes, but nightly accepts them if you use `#![feature(attr_literals)]`.

File: src/primval.rs (hunk 2 above)
Comment (solson) [inline src/primval.rs]: Since there's no existing method mapping `BinOp`s to their symbols, I'll just merge this as-is.

Comment (oli-obk) [inline src/primval.rs]: but there is!

Comment (solson) [inline src/primval.rs]: Or not! I didn't notice `as_hir_binop`.

File: src/primval.rs [hunk 3]
```diff
@@ -49,8 +49,7 @@ pub fn binary_op<'tcx>(bin_op: mir::BinOp, left: PrimVal, right: PrimVal) -> Eva
                 BitOr  => $v($l | $r),
 
                 // these have already been handled
-                Shl => bug!("`<<` operation should already have been handled"),
-                Shr => bug!("`>>` operation should already have been handled"),
+                Shl | Shr => bug!("`bin_op` operation should already have been handled", bin_op.to_hir_binop().as_str()),
```
Comment (solson) [inline src/primval.rs]: `s/bin_op/{}/`

**Pull request merged**